Fatal crash - NullPointerException: Attempt to invoke virtual method 'int java.lang.String.length()' on a null object reference

- I have searched the history issue list, and have not found a similar bug. 

- The bug was found in the version 2.97, and were reproduciable on both a Google Plxel 3 and an Android emulator 6.0 device

- Repro Steps: 

I have the event trace that leads to the crash point, but I have not succeeded to reproduce this. The crash stack indicates this issue related to a background thread, which may explain why it is not easy to reproduce.

Let me know if you need from info from my side, e..g., the event trace. 

-Crash stack (from the latest release version 2.97 (2020-03-23 21:22) )

```
 FATAL EXCEPTION: AsyncTask #2
 Process: free.rm.skytube.extra, PID: 21401
 java.lang.RuntimeException: An error occurred while executing doInBackground()
       at android.os.AsyncTask$3.done(AsyncTask.java:309)
       at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:354)
       at java.util.concurrent.FutureTask.setException(FutureTask.java:223)
       at java.util.concurrent.FutureTask.run(FutureTask.java:242)
       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1113)
       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:588)
       at java.lang.Thread.run(Thread.java:818)
 Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'int java.lang.String.length()' on a null object reference
       at java.io.StringReader.<init>(StringReader.java:47)
       at org.jsoup.parser.Parser.parse(Parser.java:107)
       at org.jsoup.Jsoup.parse(Jsoup.java:31)
       at org.schabi.newpipe.extractor.services.youtube.extractors.YoutubeChannelExtractor.getPage(YoutubeChannelExtractor.java:193)
       at free.rm.skytube.businessobjects.YouTube.newpipe.Pager.getNextPage(Pager.java:80)
       at free.rm.skytube.businessobjects.YouTube.NewPipeVideos.getNextVideos(NewPipeVideos.java:56)
       at free.rm.skytube.businessobjects.YouTube.Tasks.GetYouTubeVideosTask.getNextVideos(GetYouTubeVideosTask.java:117)
       at free.rm.skytube.businessobjects.YouTube.Tasks.GetYouTubeVideosTask.doInBackground(GetYouTubeVideosTask.java:94)
       at free.rm.skytube.businessobjects.YouTube.Tasks.GetYouTubeVideosTask.doInBackground(GetYouTubeVideosTask.java:39)
       at android.os.AsyncTask$2.call(AsyncTask.java:295)
       at java.util.concurrent.FutureTask.run(FutureTask.java:237)
       ... 3 more
```

--- a/app/src/main/java/free/rm/skytube/businessobjects/YouTube/NewPipeChannelVideos.java
+++ b/app/src/main/java/free/rm/skytube/businessobjects/YouTube/NewPipeChannelVideos.java
@@ -16,12 +16,10 @@
  */
  package free.rm.skytube.businessobjects.YouTube;
 
-import java.io.IOException;
-
-import org.schabi.newpipe.extractor.exceptions.ExtractionException;
 import org.schabi.newpipe.extractor.stream.StreamInfoItem;
 
 import free.rm.skytube.app.Utils;
+import free.rm.skytube.businessobjects.YouTube.newpipe.NewPipeException;
 import free.rm.skytube.businessobjects.YouTube.newpipe.NewPipeService;
 import free.rm.skytube.businessobjects.YouTube.newpipe.VideoPager;
 
@@ -51,7 +49,7 @@ public class NewPipeChannelVideos extends NewPipeVideos<StreamInfoItem> implemen
     }
 
     @Override
-    protected VideoPager createNewPager() throws ExtractionException, IOException {
+    protected VideoPager createNewPager() throws NewPipeException {
         Utils.requireNonNull(channelId, "channelId missing");
         return NewPipeService.get().getChannelPager(channelId);
     }

--- a/app/src/main/java/free/rm/skytube/businessobjects/YouTube/NewPipeVideoBySearch.java
+++ b/app/src/main/java/free/rm/skytube/businessobjects/YouTube/NewPipeVideoBySearch.java
@@ -16,11 +16,9 @@
  */
 package free.rm.skytube.businessobjects.YouTube;
 
-import java.io.IOException;
-
 import org.schabi.newpipe.extractor.InfoItem;
-import org.schabi.newpipe.extractor.exceptions.ExtractionException;
 
+import free.rm.skytube.businessobjects.YouTube.newpipe.NewPipeException;
 import free.rm.skytube.businessobjects.YouTube.newpipe.NewPipeService;
 import free.rm.skytube.businessobjects.YouTube.newpipe.VideoPager;
 
@@ -34,7 +32,7 @@ public class NewPipeVideoBySearch extends NewPipeVideos<InfoItem> {
     }
 
     @Override
-    protected VideoPager createNewPager() throws ExtractionException, IOException {
+    protected VideoPager createNewPager() throws NewPipeException {
         return NewPipeService.get().getSearchResult(query);
     }
 

--- a/app/src/main/java/free/rm/skytube/businessobjects/YouTube/NewPipeVideos.java
+++ b/app/src/main/java/free/rm/skytube/businessobjects/YouTube/NewPipeVideos.java
@@ -26,6 +26,7 @@ import org.schabi.newpipe.extractor.exceptions.ExtractionException;
 import free.rm.skytube.businessobjects.Logger;
 import free.rm.skytube.businessobjects.YouTube.POJOs.CardData;
 import free.rm.skytube.businessobjects.YouTube.POJOs.YouTubeVideo;
+import free.rm.skytube.businessobjects.YouTube.newpipe.NewPipeException;
 import free.rm.skytube.businessobjects.YouTube.newpipe.VideoPager;
 
 /**
@@ -37,10 +38,10 @@ public abstract class NewPipeVideos<ITEM extends InfoItem> extends GetYouTubeVid
 
     private VideoPager pager;
 
-    protected abstract VideoPager createNewPager() throws ExtractionException, IOException;
+    protected abstract VideoPager createNewPager() throws NewPipeException;
 
     @Override
-    public void init() throws IOException {
+    public void init() {
     }
 
     @Override
@@ -48,7 +49,7 @@ public abstract class NewPipeVideos<ITEM extends InfoItem> extends GetYouTubeVid
         if (pager == null) {
             try {
                 pager = createNewPager();
-            } catch (ExtractionException | IOException e) {
+            } catch (Exception e) {
                 Logger.e(this, "An error has occurred while getting videos:" + e.getMessage(), e);
                 setLastException(e);
                 return Collections.emptyList();
@@ -56,7 +57,7 @@ public abstract class NewPipeVideos<ITEM extends InfoItem> extends GetYouTubeVid
         }
         try {
             return pager.getNextPage();
-        } catch (IOException | ExtractionException e) {
+        } catch (Exception e) {
             Logger.e(this, "An error has occurred while getting videos:" + e.getMessage(), e);
             setLastException(e);
             return Collections.emptyList();

--- a/app/src/main/java/free/rm/skytube/businessobjects/YouTube/Tasks/GetBulkSubscriptionVideosTask.java
+++ b/app/src/main/java/free/rm/skytube/businessobjects/YouTube/Tasks/GetBulkSubscriptionVideosTask.java
@@ -34,6 +34,7 @@ import free.rm.skytube.businessobjects.AsyncTaskParallel;
 import free.rm.skytube.businessobjects.Logger;
 import free.rm.skytube.businessobjects.YouTube.POJOs.YouTubeChannel;
 import free.rm.skytube.businessobjects.YouTube.POJOs.YouTubeVideo;
+import free.rm.skytube.businessobjects.YouTube.newpipe.NewPipeException;
 import free.rm.skytube.businessobjects.YouTube.newpipe.NewPipeService;
 import free.rm.skytube.businessobjects.db.SubscriptionsDb;
 import free.rm.skytube.businessobjects.utils.Predicate;
@@ -166,9 +167,8 @@ public class GetBulkSubscriptionVideosTask extends AsyncTaskParallel<Void, GetBu
             // assume, they are older, and already seen
             predicate.removeIf(videos);
             return videos;
-        } catch (ExtractionException | IOException e) {
+        } catch (NewPipeException e) {
             Logger.e(this, "Error during fetching channel page for " + channelId + ",msg:" + e.getMessage(), e);
-            e.printStackTrace();
             return Collections.EMPTY_LIST;
         }
     }

--- a/app/src/main/java/free/rm/skytube/businessobjects/YouTube/newpipe/CommentPager.java
+++ b/app/src/main/java/free/rm/skytube/businessobjects/YouTube/newpipe/CommentPager.java
@@ -19,7 +19,6 @@ package free.rm.skytube.businessobjects.YouTube.newpipe;
 import org.schabi.newpipe.extractor.ListExtractor;
 import org.schabi.newpipe.extractor.StreamingService;
 import org.schabi.newpipe.extractor.comments.CommentsInfoItem;
-import org.schabi.newpipe.extractor.exceptions.ParsingException;
 
 import java.util.ArrayList;
 import java.util.List;
@@ -33,7 +32,7 @@ public class CommentPager extends Pager<CommentsInfoItem, YouTubeCommentThread> 
     }
 
     @Override
-    protected List<YouTubeCommentThread> extract(ListExtractor.InfoItemsPage<CommentsInfoItem> page) throws ParsingException {
+    protected List<YouTubeCommentThread> extract(ListExtractor.InfoItemsPage<CommentsInfoItem> page) {
         List<YouTubeCommentThread> result = new ArrayList<>(page.getItems().size());
         for (CommentsInfoItem infoItem : page.getItems()) {
             YouTubeCommentThread thread = new YouTubeCommentThread(new YouTubeComment(

--- a/app/src/main/java/free/rm/skytube/businessobjects/YouTube/newpipe/NewPipeException.java
+++ b/app/src/main/java/free/rm/skytube/businessobjects/YouTube/newpipe/NewPipeException.java
@@ -1,0 +1,27 @@
+/*
+ * SkyTube
+ * Copyright (C) 2020  Zsombor Gegesy
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation (version 3 of the License).
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package free.rm.skytube.businessobjects.YouTube.newpipe;
+
+public class NewPipeException extends Exception {
+    public NewPipeException(String message) {
+        super(message);
+    }
+
+    public NewPipeException(String message, Throwable cause) {
+        super(message, cause);
+    }
+}

--- a/app/src/main/java/free/rm/skytube/businessobjects/YouTube/newpipe/NewPipeService.java
+++ b/app/src/main/java/free/rm/skytube/businessobjects/YouTube/newpipe/NewPipeService.java
@@ -126,7 +126,7 @@ public class NewPipeService {
      * @throws ExtractionException
      * @throws IOException
      */
-    private List<YouTubeVideo> getChannelVideos(String channelId) throws ExtractionException, IOException {
+    private List<YouTubeVideo> getChannelVideos(String channelId) throws NewPipeException {
         VideoPager pager = getChannelPager(channelId);
         List<YouTubeVideo> result = pager.getNextPageAsVideos();
         Logger.i(this, "getChannelVideos for %s(%s)  -> %s videos", pager.getChannel().getTitle(), channelId, result.size());
@@ -140,7 +140,7 @@ public class NewPipeService {
      * @throws ExtractionException
      * @throws IOException
      */
-    private List<YouTubeVideo> getFeedVideos(String channelId) throws ExtractionException, IOException {
+    private List<YouTubeVideo> getFeedVideos(String channelId) throws ExtractionException, IOException, NewPipeException {
         final String url = getListLinkHandler(channelId).getUrl();
         final FeedExtractor feedExtractor = streamingService.getFeedExtractor(url);
         if (feedExtractor == null) {
@@ -159,36 +159,48 @@ public class NewPipeService {
      * @throws ExtractionException
      * @throws IOException
      */
-    public List<YouTubeVideo> getVideosFromFeedOrFromChannel(String channelId) throws IOException, ExtractionException {
+    public List<YouTubeVideo> getVideosFromFeedOrFromChannel(String channelId) throws NewPipeException {
         try {
             List<YouTubeVideo> videos = getFeedVideos(channelId);
             if (videos != null) {
                 return videos;
             }
-        } catch (IOException | ExtractionException e) {
+        } catch (IOException | ExtractionException | RuntimeException | NewPipeException e) {
             Logger.e(this, "Unable to get videos from a feed " + channelId + " : "+ e.getMessage(), e);
         }
         return getChannelVideos(channelId);
     }
 
-    public VideoPager getChannelPager(String channelId) throws ExtractionException, IOException {
-        ChannelExtractor channelExtractor = getChannelExtractor(channelId);
+    public VideoPager getChannelPager(String channelId) throws NewPipeException {
+        try {
+            ChannelExtractor channelExtractor = getChannelExtractor(channelId);
 
-        YouTubeChannel channel = createInternalChannel(channelExtractor);
-        return new VideoPager(streamingService, (ListExtractor) channelExtractor, channel);
+            YouTubeChannel channel = createInternalChannel(channelExtractor);
+            return new VideoPager(streamingService, (ListExtractor) channelExtractor, channel);
+        } catch (ExtractionException | IOException | RuntimeException e) {
+            throw new NewPipeException("Getting videos for " + channelId + " fails:" + e.getMessage(), e);
+        }
     }
 
-    public PlaylistPager getPlaylistPager(String channelId) throws ExtractionException, IOException {
-        ListLinkHandler channelList = getListLinkHandler(channelId);
-        PlaylistExtractor playlistExtractor = streamingService.getPlaylistExtractor(channelList);
-        playlistExtractor.fetchPage();
-        return new PlaylistPager(streamingService, playlistExtractor);
+    public PlaylistPager getPlaylistPager(String channelId) throws NewPipeException {
+        try {
+            ListLinkHandler channelList = getListLinkHandler(channelId);
+            PlaylistExtractor playlistExtractor = streamingService.getPlaylistExtractor(channelList);
+            playlistExtractor.fetchPage();
+            return new PlaylistPager(streamingService, playlistExtractor);
+        } catch (ExtractionException | IOException | RuntimeException e) {
+            throw new NewPipeException("Getting playlists for " + channelId + " fails:" + e.getMessage(), e);
+        }
     }
 
-    public CommentPager getCommentPager(String videoId) throws ExtractionException {
-        final ListLinkHandler linkHandler = streamingService.getCommentsLHFactory().fromId(videoId);
-        final CommentsExtractor commentsExtractor = streamingService.getCommentsExtractor(linkHandler);
-        return new CommentPager(streamingService, commentsExtractor);
+    public CommentPager getCommentPager(String videoId) throws NewPipeException {
+        try {
+            final ListLinkHandler linkHandler = streamingService.getCommentsLHFactory().fromId(videoId);
+            final CommentsExtractor commentsExtractor = streamingService.getCommentsExtractor(linkHandler);
+            return new CommentPager(streamingService, commentsExtractor);
+        } catch (ExtractionException | RuntimeException e) {
+            throw new NewPipeException("Getting comments for " + videoId + " fails:" + e.getMessage(), e);
+        }
     }
 
     /**
@@ -198,7 +210,7 @@ public class NewPipeService {
      * @throws ExtractionException
      * @throws IOException
      */
-    public YouTubeChannel getChannelDetails(String channelId) throws ExtractionException, IOException {
+    public YouTubeChannel getChannelDetails(String channelId) throws NewPipeException {
         Utils.requireNonNull(channelId, "channelId");
         VideoPager pager = getChannelPager(channelId);
         // get the channel, and add all the videos from the first page
@@ -343,10 +355,14 @@ public class NewPipeService {
         return result;
     }
 
-    public VideoPager getSearchResult(String query) throws ExtractionException, IOException {
-        SearchExtractor extractor = streamingService.getSearchExtractor(query);
-        extractor.fetchPage();
-        return new VideoPager(streamingService, extractor, null);
+    public VideoPager getSearchResult(String query) throws NewPipeException {
+        try {
+            SearchExtractor extractor = streamingService.getSearchExtractor(query);
+            extractor.fetchPage();
+            return new VideoPager(streamingService, extractor, null);
+        } catch (ExtractionException | IOException | RuntimeException e) {
+            throw new NewPipeException("Getting search result for " + query + " fails:" + e.getMessage(), e);
+        }
     }
 
     /**

--- a/app/src/main/java/free/rm/skytube/businessobjects/YouTube/newpipe/Pager.java
+++ b/app/src/main/java/free/rm/skytube/businessobjects/YouTube/newpipe/Pager.java
@@ -78,14 +78,18 @@ public abstract class Pager<I extends InfoItem, O> implements PagerBackend<O> {
      * @throws IOException
      * @throws ExtractionException
      */
-    public List<O> getNextPage() throws ParsingException, IOException, ExtractionException {
+    public List<O> getNextPage() throws NewPipeException {
         if (!hasNextPage) {
             return Collections.emptyList();
         }
-        if (nextPageUrl == null || nextPageUrl.isEmpty()) {
-            return process(channelExtractor.getInitialPage());
-        } else {
-            return process(channelExtractor.getPage(nextPageUrl));
+        try {
+            if (nextPageUrl == null || nextPageUrl.isEmpty()) {
+                return process(channelExtractor.getInitialPage());
+            } else {
+                return process(channelExtractor.getPage(nextPageUrl));
+            }
+        } catch (IOException| ExtractionException| RuntimeException e) {
+            throw new NewPipeException("getNextPage on "+this+", nextPageUrl="+nextPageUrl, e);
         }
     }
 
@@ -93,20 +97,20 @@ public abstract class Pager<I extends InfoItem, O> implements PagerBackend<O> {
     public List<O> getSafeNextPage() {
         try {
             return getNextPage();
-        } catch (Exception e) {
+        } catch (NewPipeException e) {
             lastException = e;
             Logger.e(this.getClass().getSimpleName(), "Error: " + e.getMessage(), e);
             return Collections.emptyList();
         }
     }
 
-    protected List<O> process(ListExtractor.InfoItemsPage<I> page) throws ParsingException {
+    protected List<O> process(ListExtractor.InfoItemsPage<I> page) throws NewPipeException {
         nextPageUrl = page.getNextPageUrl();
         hasNextPage = nextPageUrl != null && !nextPageUrl.isEmpty();
         return extract(page);
     }
 
 
-    protected abstract List<O> extract(InfoItemsPage<I> page) throws ParsingException;
+    protected abstract List<O> extract(InfoItemsPage<I> page) throws NewPipeException;
 
 }

--- a/app/src/main/java/free/rm/skytube/businessobjects/YouTube/newpipe/PlaylistPager.java
+++ b/app/src/main/java/free/rm/skytube/businessobjects/YouTube/newpipe/PlaylistPager.java
@@ -2,7 +2,6 @@ package free.rm.skytube.businessobjects.YouTube.newpipe;
 
 import org.schabi.newpipe.extractor.ListExtractor;
 import org.schabi.newpipe.extractor.StreamingService;
-import org.schabi.newpipe.extractor.exceptions.ParsingException;
 import org.schabi.newpipe.extractor.localization.DateWrapper;
 import org.schabi.newpipe.extractor.playlist.PlaylistExtractor;
 import org.schabi.newpipe.extractor.stream.StreamInfoItem;
@@ -19,7 +18,7 @@ class PlaylistPager extends Pager<StreamInfoItem, YouTubePlaylist>  {
     }
 
     @Override
-    protected List<YouTubePlaylist> extract(ListExtractor.InfoItemsPage<StreamInfoItem> page) throws ParsingException {
+    protected List<YouTubePlaylist> extract(ListExtractor.InfoItemsPage<StreamInfoItem> page) {
         List<YouTubePlaylist> result = new ArrayList<>(page.getItems().size());
         Logger.i(this, "extract from %s, items: %s", page, page.getItems().size());
         for (StreamInfoItem item: page.getItems()) {

--- a/app/src/main/java/free/rm/skytube/businessobjects/YouTube/newpipe/VideoPager.java
+++ b/app/src/main/java/free/rm/skytube/businessobjects/YouTube/newpipe/VideoPager.java
@@ -22,6 +22,8 @@ import org.schabi.newpipe.extractor.StreamingService;
 import org.schabi.newpipe.extractor.channel.ChannelInfoItem;
 import org.schabi.newpipe.extractor.exceptions.ExtractionException;
 import org.schabi.newpipe.extractor.exceptions.ParsingException;
+import org.schabi.newpipe.extractor.linkhandler.LinkHandler;
+import org.schabi.newpipe.extractor.linkhandler.LinkHandlerFactory;
 import org.schabi.newpipe.extractor.playlist.PlaylistInfoItem;
 import org.schabi.newpipe.extractor.stream.StreamInfoItem;
 
@@ -51,7 +53,7 @@ public class VideoPager extends Pager<InfoItem, CardData> {
     }
 
     @Override
-    protected List<CardData> extract(ListExtractor.InfoItemsPage<InfoItem> page) throws ParsingException {
+    protected List<CardData> extract(ListExtractor.InfoItemsPage<InfoItem> page) throws NewPipeException {
         List<CardData> result = new ArrayList<>(page.getItems().size());
         Logger.i(this, "extract from %s, items: %s", page, page.getItems().size());
         int repeatCounter = 0;
@@ -59,7 +61,7 @@ public class VideoPager extends Pager<InfoItem, CardData> {
 
         for (InfoItem infoItem : page.getItems()) {
             if (infoItem instanceof StreamInfoItem) {
-                String id = streamLinkHandler.getId(infoItem.getUrl());
+                String id = getId(streamLinkHandler, infoItem.getUrl());
                 StreamInfoItem streamInfo = (StreamInfoItem) infoItem;
                 if (seenVideos.contains(id)) {
                     repeatCounter++;
@@ -69,7 +71,7 @@ public class VideoPager extends Pager<InfoItem, CardData> {
                 }
             } else if (infoItem instanceof PlaylistInfoItem) {
                 PlaylistInfoItem playlistInfoItem = (PlaylistInfoItem) infoItem;
-                result.add(convert(playlistInfoItem, playlistLinkHandler.getId(infoItem.getUrl())));
+                result.add(convert(playlistInfoItem, getId(playlistLinkHandler, infoItem.getUrl())));
             } else if (infoItem instanceof ChannelInfoItem) {
                 ChannelInfoItem channelInfoItem = (ChannelInfoItem) infoItem;
                 result.add(convert(channelInfoItem));
@@ -82,7 +84,15 @@ public class VideoPager extends Pager<InfoItem, CardData> {
         return result;
     }
 
-    public List<YouTubeVideo> getNextPageAsVideos() throws IOException, ExtractionException {
+    private String getId(LinkHandlerFactory handler, String url) throws NewPipeException {
+        try {
+            return handler.getId(url);
+        } catch (ParsingException e) {
+            throw new NewPipeException("Unable to convert " + url + " with " + handler, e);
+        }
+    }
+
+    public List<YouTubeVideo> getNextPageAsVideos() throws NewPipeException {
         List<CardData> cards = getNextPage();
         List<YouTubeVideo> result = new ArrayList<>(cards.size());
         for (CardData cardData: cards) {

--- a/app/src/main/java/free/rm/skytube/businessobjects/db/Tasks/GetChannelInfo.java
+++ b/app/src/main/java/free/rm/skytube/businessobjects/db/Tasks/GetChannelInfo.java
@@ -30,6 +30,7 @@ import free.rm.skytube.businessobjects.AsyncTaskParallel;
 import free.rm.skytube.businessobjects.Logger;
 import free.rm.skytube.businessobjects.YouTube.POJOs.YouTubeChannel;
 import free.rm.skytube.businessobjects.YouTube.POJOs.YouTubeChannelInterface;
+import free.rm.skytube.businessobjects.YouTube.newpipe.NewPipeException;
 import free.rm.skytube.businessobjects.YouTube.newpipe.NewPipeService;
 import free.rm.skytube.businessobjects.db.SubscriptionsDb;
 
@@ -80,7 +81,7 @@ public class GetChannelInfo extends AsyncTaskParallel<String, Void, YouTubeChann
             try {
                 channel = NewPipeService.get().getChannelDetails(channelId);
                 db.cacheChannel(channel);
-            } catch (ExtractionException | IOException e) {
+            } catch (NewPipeException | RuntimeException e) {
                 exception = e;
             }
         }

